KOGITO-8624: Provide autocomplete support for JQ expressions

AjayJagan opened the pull request:
Provide support for jq expressions

Related JIRA's:
 - https://issues.redhat.com/browse/KOGITO-8624
 - https://issues.redhat.com/browse/KOGITO-8625

https://user-images.githubusercontent.com/36824134/223388812-9815677c-16b7-4c65-9a6a-dbfff4d7f04b.mp4

fantonangeli commented on packages/serverless-workflow-jq-expressions/src/channel/impl/JqExpressionReadSchemasImpl.ts:
```diff
@@ -0,0 +1,88 @@
+} from "@kie-tools/serverless-workflow-service-catalog/dist/api";
+import { parseApiContent } from "@kie-tools/serverless-workflow-service-catalog/dist/channel";
+import fetch from "cross-fetch";
+export class JqExpressionReadSchemasImpl implements JqExpressionReadSchemas {
```
I didn't see any channel communication with `serverless-workflow-jq-expressions`, am I wrong?
In case there is no channel communication, I think this class can be treated as a normal class, just moving this file in a different dir under `/src`. WDYT?

fantonangeli commented on packages/serverless-workflow-service-catalog/src/channel/parsers/impl/jsonschema/jsonschema.ts:
```diff
@@ -0,0 +1,85 @@
+}
+
+function resolveArgumentType(type: string): SwfServiceCatalogFunctionArgumentType {
+  switch (type) {
+    case "boolean":
+      return SwfServiceCatalogFunctionArgumentType.boolean;
+    case "object":
+      return SwfServiceCatalogFunctionArgumentType.object;
+    case "number":
+      return SwfServiceCatalogFunctionArgumentType.number;
+    case "string":
+      return SwfServiceCatalogFunctionArgumentType.string;
+    case "integer":
+      return SwfServiceCatalogFunctionArgumentType.integer;
+    case "array":
+      return SwfServiceCatalogFunctionArgumentType.array;
+  }
+  return SwfServiceCatalogFunctionArgumentType.object;
```
This suggestion needs to be tried as I'm not 100% sure.

```suggestion
  return SwfServiceCatalogFunctionArgumentType[type as keyof typeof SwfServiceCatalogFunctionArgumentType];
```

fantonangeli reviewed (changes requested):
Hi @AjayJagan ,
I'm amazed by your work and how fast you started extending the Code Completion.
This feature will improve a lot the user experience.

I wrote some comments about your code, but they are mostly minor modifications.
Together with those comments, I spotted a bug:
- when trying to complete after a whitespace, the tools shows completions from the schema. eg:
`"operation": ". :dart: "

AjayJagan commented:
Hey @fantonangeli, Thanks a lot for your review :) I have resolved your comments now. Let me know if they are good enough.

fantonangeli commented on packages/vscode-extension-serverless-workflow-editor/src/extension/languageService/VsCodeSwfLanguageService.ts:
```diff
@@ -143,20 +169,25 @@ export class VsCodeSwfLanguageService {
 
+  private getSchemaFilePosixPath(args: { doc: TextDocument; schemaPath: string }) {
+    const baseFileAbsolutePosixPath = vscode.Uri.parse(args.doc.uri).path;
+    return baseFileAbsolutePosixPath
+      .split("/")
+      .splice(0, baseFileAbsolutePosixPath.split("/").length - 1)
+      .join("/")
```
Can you rewrite this part storing `baseFileAbsolutePosixPath.split("/")` in a const or even better with a regex?
Maybe this can help https://regex101.com/r/VCx0M7/1

AjayJagan replied:
fixed this one as well and thanks a lot for the regex :)

fantonangeli reviewed (changes requested):
Thank you @AjayJagan for fixing my previous comments.
Here is just one small improvement

AjayJagan commented:
hey @fantonangeli , I have used regex is all the places to capture the word. I have checked and everything seems to work but  I would like you too break it :D so that no edge cases are missed

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -129,6 +138,190 @@ function getStateNameCompletion(
+          completion: filteredName,
+          kind: CompletionItemKind.Function,
+          label: filteredName,
+          filterText: args.wordToSearch.length ? args.wordToSearch : filteredName,
```
Together with my previous comment

```suggestion
          filterText: isWordToSearchEmpty ? args.wordToSearch : filteredName,
```

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -129,6 +138,190 @@ function getStateNameCompletion(
+          label: filteredName,
+          filterText: args.wordToSearch.length ? args.wordToSearch : filteredName,
+          detail: "Reusable functions(expressions) defined in the functions array",
+          overwriteRange: Range.create(
```
`overwriteRange` never changes. 
Can you declare it as a constant outside of the `.map()` cycle?

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -262,28 +266,54 @@ async function getJqFunctionCompletions(
+    } else {
+      wordToSearch = inputVariableMatch[1];
+    }
     return await getJqInputVariablesCompletions({ ...args, wordToSearch });
```
```suggestion
    return await getJqInputVariablesCompletions({ ...args, wordToSearch: (inputVariableMatch[1] === undefined)?".":inputVariableMatch[1] });
```

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -262,28 +266,54 @@ async function getJqFunctionCompletions(
+    } else {
+      wordToSearch = reusableFunctionMatch[1];
+    }
     return getReusableFunctionCompletion({ ...args, wordToSearch });
```
Same as previous suggestion

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -262,28 +266,54 @@ async function getJqFunctionCompletions(
+  }
   const result = jqBuiltInFunctions
-    .filter((func: { functionName: string; description: string }) => func.functionName.startsWith(wordToSearch))
+    .filter((func: { functionName: string; description: string }) => {
```
Same as the comment to move the `wordToSearch.lenght` check outside of the cycle

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -262,28 +266,54 @@ async function getJqFunctionCompletions(
         label: filteredFunc.functionName,
         detail: filteredFunc.description,
-        filterText: wordToSearch,
+        filterText: wordToSearch.length ? wordToSearch : filteredFunc.functionName,
```
I would use the same constant as for the previous comment

fantonangeli commented on packages/serverless-workflow-language-service/src/channel/SwfLanguageServiceCodeCompletion.ts:
```diff
@@ -262,28 +266,54 @@ async function getJqFunctionCompletions(
+    }
     return getReusableFunctionCompletion({ ...args, wordToSearch });
   }
+  const removeSpecialChar = slicedValue.replace(/[^a-zA-Z _()]/g, "");
+  const builtInFunctionMatch = removeSpecialChar.match(/\s(\w+)?$/);
+  let wordToSearch = "";
+  if (builtInFunctionMatch === null) {
+    if (removeSpecialChar.length) {
+      wordToSearch = removeSpecialChar.trim();
+    }
+  } else if (builtInFunctionMatch[1] === undefined) {
+    wordToSearch = "";
+  } else if (builtInFunctionMatch[1].length) {
+    wordToSearch = builtInFunctionMatch[1];
+  }
```
I think moving this code into a function and declaring here `wordToSearch` as a const will make the code easier to read and test. WDYT?

caponetto commented:
@AjayJagan consider moving this PR to "ready to review" so the CI build can run.

fantonangeli commented on packages/serverless-workflow-language-service/tests/SwfJsonLanguageService.test.ts:
```diff
@@ -697,7 +721,338 @@ describe("SWF LS JSON", () => {
+            false
+          );
+          expect(completionItems.length).toMatchSnapshot();
+          if (completionItems.length > 10) {
+            expect(completionItems.slice(-5)).toMatchSnapshot();
+          } else {
+            expect(completionItems).toMatchSnapshot();
+          }
```
This should be enough to test only the first 5 elements.
This can be applied to other Jq tests.

```suggestion
          expect(completionItems.slice(0,5)).toMatchSnapshot();
```

fantonangeli commented on packages/serverless-workflow-language-service/tests/SwfYamlLanguageService.test.ts:
```diff
@@ -787,7 +815,234 @@ functions:
       });
     });
+    describe.each([["Jq operation completion for built-in functions"], ["Jq operation completion for variables"]])(
+      `%s completion`,
```
Please apply the same comments of `SwfJsonLanguageService.test.ts`

fantonangeli reviewed (changes requested):
I share a partial review here.
Thanks again for writing so many tests.

fantonangeli reviewed (changes requested):
Below are a couple of very small comments,  then everything should be ready.

fantonangeli commented:
Hi @AjayJagan,
To let the completion be triggered when the user press "." character, I would evaluate changing the line:
packages/serverless-workflow-text-editor/src/editor/textEditor/augmentation/completion/index.ts:31
adding the "." char:
`triggerCharacters: [" ", ":", '"', "."],`

WDYT?
Strangely this seems to work well for Yaml, but not for Json.
I would also reflect this to the VSCode Extension.



- For Yaml, I noticed that the code below doesn't trigger the right completions :
functions:
  - name: max
    type: expression
    operation: .:dart:

caponetto commented:
@AjayJagan Can you please add a short gif/video to the PR description to show off this new feature?

tomasdavidorg commented:
@AjayJagan I am unable to use autocompletion after '.'. After I added '"dataInputSchema": "schema/someSchema.json"', I don't see any suggestion from the schema. Do I have to setup anything else?

fantonangeli commented:
> @AjayJagan I am unable to use autocompletion after `.`. After I added `"dataInputSchema": "schema/someSchema.json"`, I don't see any suggestion from the schema. Do I have to setup anything else?

Hi @tomasdavidorg, do you mean typing a `.` you don't see autocompletion automatically open?
Have you tried with Ctrl-space?

AjayJagan commented on packages/serverless-workflow-jq-expressions/src/utils/JqBuiltInFunctions.ts:
```diff
@@ -0,0 +1,584 @@
+      "The walk(f) function applies f recursively to every component of the input entity. When an array is encountered, f is first applied to its elements and then to the array itself; when an object is encountered, f is first applied to all the values and then to the object. In practice, f will usually test the type of its input, as illustrated in the following examples. The first example highli ...
+  },
+  {
+    functionName: "$ENV",
```
I guess the `$` is the problem here.. not sure why it does not add anything

AjayJagan commented on packages/serverless-workflow-jq-expressions/src/utils/JqBuiltInFunctions.ts:
```diff
@@ -0,0 +1,584 @@
+      "Stops the jq program with no further outputs. The input will be printed on stderr as raw output (i.e., strings will not have double quotes) with no decoration, not even a newline.",
+  },
+  {
+    functionName: "$__loc__",
```
same here

AjayJagan commented on packages/serverless-workflow-jq-expressions/src/utils/JqBuiltInFunctions.ts:
```diff
@@ -0,0 +1,584 @@
+ */
+
+// The below mentioned functions are referenced from https://stedolan.github.io/jq/manual/.
+export const jqBuiltInFunctions = [
```
by math functions you mean the `+`, `-` like that? 
What should be in this list, all the keywords? -> the idea is to have all the keywords in the list .. I tried to gather as much as I can from here:https://stedolan.github.io/jq/manual. But we need more .. we can always add :)

tomasdavidorg replied:
I meant https://stedolan.github.io/jq/manual/#Math ... `sin`, `cos`, etc., otherwise I think they should be all. I will check it more.

AjayJagan replied:
yea so the doc was a bit big hehe.. So I might have missed a few things there.. so let me know so that I can add them 👍

AjayJagan commented:
> @AjayJagan I am unable to use autocompletion after `.`. After I added `"dataInputSchema": "schema/someSchema.json"`, I don't see any suggestion from the schema. Do I have to setup anything else?

So is this happening in the vscode or the webtools. The 

> @AjayJagan I am unable to use autocompletion after `.`. After I added `"dataInputSchema": "schema/someSchema.json"`, I don't see any suggestion from the schema. Do I have to setup anything else?

so the for it to work we need to put a jsonschema in the `schema` dir inside the resources folder of the project(something like this: https://github.com/kiegroup/kogito-examples/blob/stable/serverless-workflow-examples/serverless-workflow-expression-quarkus/src/main/resources/schema/expression.json)
Apart from that in the webtools we need to use `ctrl+space` to trigger the completion

tomasdavidorg commented:
Yes `ctrl + space`. I will try once more. Thanks for the answers 🙂

tomasdavidorg commented:
@AjayJagan I finished my review. I reported one improvemnt jira https://issues.redhat.com/browse/KOGITO-8829

tomasdavidorg commented on packages/serverless-workflow-jq-expressions/src/utils/JqBuiltInFunctions.ts:
```diff
@@ -0,0 +1,584 @@
+  {
+    functionName: "booleans",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "numbers",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "normals",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "finites",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "strings",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "nulls",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "values",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
+  {
+    functionName: "scalars",
+    description:
+      "These built-ins select only inputs that are arrays, objects, iterables (arrays or objects), booleans, numbers, normal numbers, finite numbers, strings, null, non-null values, and non-iterables, respectively.",
+  },
```
https://issues.redhat.com/browse/KOGITO-8831

AjayJagan commented on packages/serverless-workflow-jq-expressions/env/index.js:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```
changed

AjayJagan commented on packages/serverless-workflow-service-catalog/src/channel/parsers/impl/jsonschema/jsonschema.ts:
```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates.
```
changed

tomasdavidorg reviewed (approved):
Thanks @AjayJagan. Looks good to me 🙂.

caponetto reviewed (approved):
Very good stuff here, @AjayJagan! Thank you!

caponetto commented:
@AjayJagan Conflicts showed up. Can you please fix them?

AjayJagan commented:
> @AjayJagan Conflicts showed up. Can you please fix them?

resolved the conflicts
